### ISSUE TITLE
Possible solution for failed to dial, failed to connect found on this lines ?

### ISSUE THREAD

**Issue by serbyxp** [helium/miner#1434]
This line is causing a lot of issues, that I'm not sure if its just me or a general issue with libp2p, but I always have issues with this failing to connect. But, I have noticed in the lines below 175-207 I fail most times on the first attempt, but on the retry It usually works. Not sure if that's just me or a common issue but If you can make this retry or attempt reconnect it might fix that... ( I know it will fix the issue for the 151-157 if it connects first, then tries to send the receipt) so it might work for the 175-207 ( which doesn't try to attempt to re /connect to the same "peer" or " challenger" endpoint @madninja 

https://github.com/helium/miner/blob/master/src/poc/miner_onion_server.erl#L151-L157

Like this section bellow that re tries after the false -> depending on the max retry variable

https://github.com/helium/miner/blob/master/src/poc/miner_onion_server.erl#L175-L207

**Comment by madninja** [comment]
The send_receipt block you highlight already has a retry as part of the function it s in: https://github.com/helium/miner/blob/master/src/poc/miner_onion_server.erl#L171 so this is not quite it, unfortunately.

**Comment by marcijel1** [comment]
Is it possible to edit the retry timer from 30 secs to 1~9 seconds, and extend the retry multiplier from 10x to ~100+ ?

Or edit this file is illegal?

I'm afraid that I lose some of my rewarded witnesses because of the late successful dialing. Usually my witnesses failed to dial at the 1st time, and success at the 2nd time, it's 30 seconds delay already. Also I have the problem that some of my witnesses are just trying to dial, and failing to dial 10x times, so it just can't communicate with the challenger.

So it's just not necessary to wait 10x30 seconds for nothing. My most tryed succesful and rewarded witness was 3x times trying, so 3x30=90 seconds would be enough maximum of time. This is the time limit within I would like to make as much retry as I allowed to.

I understand that these p2p/dial fail errors not just my hotspot's error, this is a system error also, but if I can, I would try to bypass from my end, and minimize the errors and fails.

**Comment by serbyxp** [comment]
well, I did a manual test, that I turned into a script. for the failed to send  witness to challenger. The reason I mentioned the " retry " on the other block is because I have no manual access to it, without compiling it myself. Their is no JRPC commands available *that I know of * for the libp2p_transport_proxy: that just starts hopping onto the next peer and so on which never retries the "stream of peers", What I am trying to say or explain is. right now its doing something like

Peer , sends receipt -> challanger . < Fails to dial > retrys, Peer,  sends receipt -> p2p/circuit/challanger, <Fails to dial>  

It doesnt always fail, but it always fails on the first attempt (for me at least). Then it moves on to the retries where it sometimes success and most time fails.What Ive found works for me manually is I look at console log, When a [ warning ] fails to dial challenger /p2p/ChallangerAddr   << I copy that address , and manually type in >> "miner peer connect /p2p/ChallangerAddr " most times it connects on the first try but always connects on the second attempt.( not all of them but 90%, the firewall relayed issue still exists, which you would replace the "miner peer connect /p2p/ChallangerAddr with the /p2p/ProxyAddr/p2p-circuit/p2p2/ChallangerAddr with known proxy / peer addresses that have successfully sent receipts to the challenger ( its peer book ). When I manually input that " miner peer connect /p2p/ChallangerAddr" after a "[ warning ] failed to dial challanger ", The next dial " successfully sends witness" 100% of the time. SO what Im Suggesting is more the Procedure of connecting...

Peer, connect to -> Challanger, Peer successfully connected, if True -> Peer -> sends receipt -> challanger, if False -> retry(3), connect to -> Challanger 

Something like that I t needs to establish a connection first before sending the receipt all willy nilly, I understand thats how P2P is supposed to work... but I mean we know the address we have to connect to so why not just establish that connection first and then send the receipt... I got tired of manually typing it in so now it just does it automaticlly and I haven't had one failed witness since. 

The failed to dial proxy server, never retries it just goes down a rabbit hole of first time failures, And I dont know of a CLI / API command to connect that, Without getting my hands dirty...




I put example below.. I'm not trying to re write a thing I have no experience in... This are the test steps that ended into a successful script.
